From 1db73e638cf535e4820fb36cb873a96e7dc7a100 Mon Sep 17 00:00:00 2001
From: Peter Chiu <chui-hao.chiu@mediatek.com>
Date: Wed, 16 Apr 2025 11:04:00 +0800
Subject: [PATCH 082/119] mtk: mt76: mt7996: fix bcn prot hw mode

Fix beacon protection hw mode.
1. Set TXD_BIP when it is beacon and beacon protection is enabled.
2. Set TXD_WLAN_IDX to BMC wcid when it is beacon and beacon
protection is enabled.
3. Update beacon content after add beacon bigtk. Remove beacon before
adding new beacon content to update the beacon TxD.

Signed-off-by: Peter Chiu <chui-hao.chiu@mediatek.com>
---
 mt76.h        |  1 +
 mt7996/mac.c  |  3 ++-
 mt7996/main.c | 11 +++++++++--
 mt7996/mcu.c  | 19 +++++++++++++++++--
 4 files changed, 29 insertions(+), 5 deletions(-)

--- a/mt76.h
+++ b/mt76.h
@@ -386,6 +386,7 @@ struct mt76_wcid {
 	u8 amsdu:1;
 	u8 phy_idx:2;
 	u8 link_id:4;
+	u8 hw_bcn_prot:1;
 	bool link_valid;
 
 	u8 rx_check_pn;
--- a/mt7996/mac.c
+++ b/mt7996/mac.c
@@ -811,7 +811,8 @@ mt7996_mac_write_txwi_80211(struct mt799
 	    info->flags & IEEE80211_TX_CTL_USE_MINRATE)
 		val |= MT_TXD1_FIXED_RATE;
 
-	if (key && multicast && ieee80211_is_robust_mgmt_frame(skb)) {
+	if ((key && multicast && ieee80211_is_robust_mgmt_frame(skb)) ||
+	    (ieee80211_is_beacon(fc) && wcid->hw_bcn_prot)) {
 		val |= MT_TXD1_BIP;
 		txwi[3] &= ~cpu_to_le32(MT_TXD3_PROTECT_FRAME);
 	}
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -186,6 +186,7 @@ mt7996_set_hw_key(struct ieee80211_hw *h
 {
 	struct mt7996_dev *dev = mt7996_hw_dev(hw);
 	int idx = key->keyidx;
+	bool is_bigtk = idx == 6 || idx == 7;
 	int err = 0;
 	unsigned int link_id;
 	unsigned long links;
@@ -236,7 +237,7 @@ mt7996_set_hw_key(struct ieee80211_hw *h
 		case WLAN_CIPHER_SUITE_BIP_CMAC_256:
 		case WLAN_CIPHER_SUITE_BIP_GMAC_128:
 		case WLAN_CIPHER_SUITE_BIP_GMAC_256:
-			if (key->keyidx == 6 || key->keyidx == 7) {
+			if (is_bigtk) {
 				wcid_keyidx = &msta_link->wcid.hw_key_idx2;
 				key->flags |= IEEE80211_KEY_FLAG_GENERATE_MMIE;
 				break;
@@ -268,10 +269,6 @@ mt7996_set_hw_key(struct ieee80211_hw *h
 			*wcid_keyidx = -1;
 		}
 
-		/* To remove BIGTK independently, FW needs an extra inband command */
-		if (cmd == DISABLE_KEY && !(idx == 6 || idx == 7))
-			continue;
-
 		mt76_wcid_key_setup(&dev->mt76, &msta_link->wcid, key);
 
 		err = mt7996_mcu_add_key(&dev->mt76, vif, key,
@@ -279,6 +276,19 @@ mt7996_set_hw_key(struct ieee80211_hw *h
 					 &msta_link->wcid, cmd);
 		if (err)
 			return err;
+
+		if (cmd == SET_KEY && is_bigtk) {
+			struct ieee80211_bss_conf *link_conf = link_conf_dereference_protected(vif,
+								    link_id);
+
+			/* Remove beacon first to update beacon Txd for beacon protection */
+			if (link_conf && link_conf->enable_beacon) {
+				link_conf->enable_beacon = 0;
+				mt7996_mcu_add_beacon(hw, vif, link_conf);
+				link_conf->enable_beacon = 1;
+				mt7996_mcu_add_beacon(hw, vif, link_conf);
+			}
+		}
 	}
 
 	return 0;
--- a/mt7996/mcu.c
+++ b/mt7996/mcu.c
@@ -2566,11 +2566,20 @@ mt7996_mcu_sta_key_tlv(struct mt76_dev *
 				sec_key->cipher_id = MCU_CIPHER_BCN_PROT_GMAC_256;
 				break;
 			case WLAN_CIPHER_SUITE_BIP_CMAC_256:
+				if (is_mt7990(dev)) {
+					sec_key->cipher_id = MCU_CIPHER_BCN_PROT_CMAC_256;
+					break;
+				}
+				fallthrough;
 			default:
 				dev_err(dev->dev, "Unsupported BIGTK cipher\n");
 				return -EOPNOTSUPP;
 			}
 			sec_key->bcn_mode = BP_SW_MODE;
+			if (is_mt7990(dev)) {
+				sec_key->bcn_mode = BP_HW_MODE;
+				wcid->hw_bcn_prot = true;
+			}
 		}
 
 		if (cipher == MCU_CIPHER_TKIP) {
@@ -2582,6 +2591,8 @@ mt7996_mcu_sta_key_tlv(struct mt76_dev *
 		/* connac3 fw use set key action to apply removing bigtk and other
 		 * group keys should just use set key to overwrite the old ones. */
 		sec->add = SET_KEY;
+		if (is_mt7990(dev) && (sec_key->key_id == 6 || sec_key->key_id == 7))
+			wcid->hw_bcn_prot = false;
 	}
 
 	return 0;
@@ -2722,11 +2733,15 @@ mt7996_mcu_beacon_cont(struct mt7996_dev
 		       struct ieee80211_bss_conf *link_conf,
 		       struct sk_buff *rskb, struct sk_buff *skb,
 		       struct bss_bcn_content_tlv *bcn,
-		       struct ieee80211_mutable_offsets *offs)
+		       struct ieee80211_mutable_offsets *offs,
+		       struct mt7996_vif_link *link)
 {
 	struct mt76_wcid *wcid = &dev->mt76.global_wcid;
 	u8 *buf;
 
+	if (link->msta_link.wcid.hw_bcn_prot)
+		wcid = &link->msta_link.wcid;
+
 	bcn->pkt_len = cpu_to_le16(MT_TXD_SIZE + skb->len);
 	bcn->tim_ie_pos = cpu_to_le16(offs->tim_offset);
 
@@ -2801,7 +2816,7 @@ int mt7996_mcu_add_beacon(struct ieee802
 	info = IEEE80211_SKB_CB(skb);
 	info->hw_queue |= FIELD_PREP(MT_TX_HW_QUEUE_PHY, mlink->band_idx);
 
-	mt7996_mcu_beacon_cont(dev, link_conf, rskb, skb, bcn, &offs);
+	mt7996_mcu_beacon_cont(dev, link_conf, rskb, skb, bcn, &offs, link);
 	if (link_conf->bssid_indicator)
 		mt7996_mcu_beacon_mbss(rskb, skb, bcn, &offs);
 	mt7996_mcu_beacon_cntdwn(rskb, skb, &offs, link_conf->csa_active);
